runner/status_rw.go

```diff
@@ -0,0 +1,120 @@
+package runner
+
+import (
+	"fmt"
+	"time"
+)
+
+// status_rw.go is how to read/write RunStatus'es
+
+// StateMask describes a set of States as a bitmask.
+type StateMask uint64
+
+// Useful StateMask constants
+const (
+	UNKNOWN_MASK    StateMask = StateMask(1 << uint(UNKNOWN))
+	PENDING_MASK              = 1 << uint(PENDING)
+	PREPARING_MASK            = 1 << uint(PREPARING)
+	RUNNING_MASK              = 1 << uint(RUNNING)
+	COMPLETE_MASK             = 1 << uint(COMPLETE)
+	FAILED_MASK               = 1 << uint(FAILED)
+	ABORTED_MASK              = 1 << uint(ABORTED)
+	TIMEDOUT_MASK             = 1 << uint(TIMEDOUT)
+	BADREQUEST_MASK           = 1 << uint(BADREQUEST)
+	DONE_MASK                 = (1<<uint(COMPLETE) |
+		1<<uint(FAILED) |
+		1<<uint(ABORTED) |
+		1<<uint(TIMEDOUT))
+	ALL_MASK = math.MaxUint64
+)
+
+// Helper Function to create StateMask that matches exactly state
+func MaskForState(state ProcessState) StateMask {
+	return 1 << uint(state)
+}
+
+// Query describes a query for RunStatuses.
+// The Runs and States are and'ed: a RunStatus matches a Query
+// if its ID is in q.Runs (or q.AllRuns) and its state is in q.States
+type Query struct {
+	Runs    []RunID   // Runs to query for
+	AllRuns bool      // Whether to match all runs
+	States  StateMask // What States to match
+}
+
+// Wait describes how to Wait.
+// It differs from StatusQuery because StatusQuery describes what RunStatus'es to match, but
+// Wait describes how to Wait
+type Wait struct {
+	// How long to wait for Statuses
+	Timeout time.Duration
+
+	// We might add whether to return as soon as one status matches, or waiting until all
+}
+
+// StatusQuerier allows reading Status by Query'ing.
+type StatusQuerier interface {
+	// Query returns all RunStatus'es matching q, waiting as described by w
+	Query(q Query, w Wait) ([]RunStatus, error)
+
+	StatusQueryNower
+}
+
+// StatusQueryNower allows Query'ing Statuses but with no Waiting
+// This is separate from StatusQuerier because talking to the Worker, e.g., we will be able to
+// QueryNow easily, but because Thrift doesn't like long waits on RPCs, it can't do
```

Comment:
How do we know thrift doesn't like long RPC waits? And do we think callers would want a raw StatusQueryNower vs the more comprehensive StatusQuerier, i.e. do we have to expose Nower as a separate iface?

And the thing about special values for runner.Wait is that I'll look to find out what those are when reading through code, but there's no corresponding flag indicating that Query() has a QueryNow() cousin.